src/shogun/lib/type_case.h

```diff
@@ -0,0 +1,369 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#ifndef SHOGUN_TYPE_CASE_H
+#define SHOGUN_TYPE_CASE_H
+
+#include <typeindex>
+#include <unordered_map>
+
+#include <shogun/lib/SGMatrix.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/lib/any.h>
+#include <shogun/lib/type_list.h>
+
+using namespace shogun;
+
+namespace shogun
+{
+	typedef Types<
+		bool, char, int8_t, uint8_t, int16_t, uint16_t, int32_t, uint32_t,
+		int64_t, uint64_t, float32_t, float64_t, floatmax_t, SGVector<int32_t>,
+		SGVector<int64_t>, SGVector<float32_t>, SGVector<float64_t>,
+		SGVector<floatmax_t>, SGMatrix<int32_t>, SGMatrix<int64_t>,
+		SGMatrix<float32_t>, SGMatrix<float64_t>, SGMatrix<floatmax_t>>::type
+		SG_TYPES;
+
+	enum class TYPE
+	{
+		PT_BOOL = 1,
+		PT_CHAR = 2,
+		PT_INT8 = 3,
+		PT_UINT8 = 4,
+		PT_INT16 = 5,
+		PT_UINT16 = 6,
+		PT_INT32 = 7,
+		PT_UINT32 = 8,
+		PT_INT64 = 9,
+		PT_UINT64 = 10,
+		PT_FLOAT32 = 11,
+		PT_FLOAT64 = 12,
+		PT_FLOATMAX = 13,
+		PT_SGOBJECT = 14,
+		PT_COMPLEX128 = 15,
+		PT_SGVECTOR_FLOAT32 = 16,
+		PT_SGVECTOR_FLOAT64 = 17,
+		PT_SGVECTOR_FLOATMAX = 18,
+		PT_SGVECTOR_INT32 = 19,
+		PT_SGVECTOR_INT64 = 20,
+		PT_SGMATRIX_FLOAT32 = 21,
+		PT_SGMATRIX_FLOAT64 = 22,
+		PT_SGMATRIX_FLOATMAX = 23,
+		PT_SGMATRIX_INT32 = 24,
+		PT_SGMATRIX_INT64 = 25,
+		PT_UNDEFINED = 26
+	};
+
+	enum class CONTAINER_TYPE
+	{
+		SCALAR = 1,
+		VECTOR = 2,
+		MATRIX = 3
```

Comment:
if there is another "container type" enum, I don't really get why we then also need to state the containers here?
Having both seems redundant...? Maybe I just don't get it

Reply:
is this used anywhere?